File: zoom/src/zoomCocoa/ZoomSavePreviewView.m
//
//  ZoomSavePreviewView.m
//  ZoomCocoa
//
//  Created by Andrew Hunter on Mon Mar 22 2004.
//  Copyright (c) 2004 Andrew Hunter. All rights reserved.
//

#import "ZoomSavePreviewView.h"
#import "ZoomSavePreview.h"

#import "ZoomClient.h"
#import "ZoomAppDelegate.h"


@implementation ZoomSavePreviewView

- (id)initWithFrame:(NSRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
		upperWindowViews = nil;
		[self setAutoresizesSubviews: YES];
		[self setAutoresizingMask: NSViewWidthSizable];
		selected = NSNotFound;
    }
    return self;
}

- (void) dealloc {
	if (upperWindowViews) [upperWindowViews release];
	
	[super dealloc];
}

- (void)drawRect:(NSRect)rect {
}

- (void) setDirectoryToUse: (NSString*) directory {
	// Get rid of our old views
	if (upperWindowViews) {
		[upperWindowViews makeObjectsPerformSelector: @selector(removeFromSuperview)];
		[upperWindowViews release];
		upperWindowViews = nil;
	}
	
	saveGamesAvailable = NO;
	upperWindowViews = [[NSMutableArray alloc] init];
	selected = NSNotFound;
	
	if (directory == nil || ![[NSFileManager defaultManager] fileExistsAtPath: directory]) {
		NSRect ourFrame = [self frame];
		ourFrame.size.height = 2;
		[self setFrame: ourFrame];
		[self setNeedsDisplay: YES];
		return;
	}
	
	// Get our frame size
	NSRect ourFrame = [self frame];
	ourFrame.size.height = 0;
	
	// Load all the zoomSave files from the given directory
	NSArray* contents = [[NSFileManager defaultManager] directoryContentsAtPath: directory];
	
	if (contents == nil) {
		return;
	}
	
	// Read in the previews from any .zoomSave packages
	NSEnumerator* fileEnum = [contents objectEnumerator];
	NSString* file;
	
	while (file = [fileEnum nextObject]) {
		if ([[[file pathExtension] lowercaseString] isEqualToString: @"zoomsave"]) {
			
			// This is a zoomSave file - load the preview
			NSString* previewFile = [directory stringByAppendingPathComponent: file];
			previewFile = [previewFile stringByAppendingPathComponent: @"ZoomPreview.dat"];
			
			BOOL isDir;
			
			if (![[NSFileManager defaultManager] fileExistsAtPath: previewFile
													  isDirectory: &isDir]) {
				// Can't be a valid zoomSave file
				continue;
			}
			
			if (isDir) {
				// Also can't be a valid zoomSave file
				continue;
			}
			
			// Presumably, this is a valid preview file...
			ZoomUpperWindow* win = [NSUnarchiver unarchiveObjectWithFile: previewFile];
			
			if (win != nil && ![win isKindOfClass: [ZoomUpperWindow class]]) continue;
			
			// We've got a valid window - add to the list of upper windows
			ZoomSavePreview* preview;
			
			preview = [[ZoomSavePreview alloc] initWithPreview: win
													  filename: previewFile];
			
			[preview setAutoresizingMask: NSViewWidthSizable];
			[preview setMenu: [self menu]];
			[self addSubview: preview];
			[upperWindowViews addObject: [preview autorelease]];
			
			saveGamesAvailable = YES;

		} else if ([[[file pathExtension] lowercaseString] isEqualToString: @"glksave"]) {
		
			// This is a glksave file
			NSString* previewFile = [directory stringByAppendingPathComponent: file];
			
			NSDictionary* previewProperties = nil;
			NSString* propertiesPath = [previewFile stringByAppendingPathComponent: @"Info.plist"];
			if ([[NSFileManager defaultManager] fileExistsAtPath: propertiesPath]) {
				previewProperties = [NSPropertyListSerialization propertyListFromData: [NSData dataWithContentsOfFile: propertiesPath]
																	 mutabilityOption: NSPropertyListImmutable
																			   format: nil
																	 errorDescription: nil];
				if (![previewProperties isKindOfClass: [NSDictionary class]]) previewProperties = nil;
			}
			
			if (!previewProperties) continue;
			
			ZoomStoryID* storyId = [[ZoomStoryID alloc] initWithIdString: [previewProperties objectForKey: @"ZoomGlkGameId"]];
			
			// Load the preview lines from the glksave directory
			NSArray* previewLines = nil;
			NSString* previewLinesPath = [previewFile stringByAppendingPathComponent: @"Preview.plist"];
			if ([[NSFileManager defaultManager] fileExistsAtPath: previewLinesPath]) {
				previewLines = [NSPropertyListSerialization propertyListFromData: [NSData dataWithContentsOfFile: previewLinesPath]
																mutabilityOption: NSPropertyListImmutable
																		  format: nil
																errorDescription: nil];
				if (![previewLines isKindOfClass: [NSArray class]]) previewLines = nil;
			}
			
			// Use some defaults if no lines are supplied
			if (!previewLines) {
				ZoomStory* story = [[NSApp delegate] findStory: storyId];
				
				if (story) {
					previewLines = [NSArray arrayWithObjects:
						[NSString stringWithFormat: @"Saved story from '%@'", [story title]],
						nil];
				} else {
					previewLines = [NSArray arrayWithObject: @"Saved story"];
				}
			}
			
			// Create the preview object
			ZoomSavePreview* preview = [[ZoomSavePreview alloc] initWithPreviewStrings: previewLines
																			  filename: propertiesPath];
			
			[preview setAutoresizingMask: NSViewWidthSizable];
			[preview setMenu: [self menu]];
			[self addSubview: preview];
			[upperWindowViews addObject: [preview autorelease]];
			
			saveGamesAvailable = YES;

		}
	}
	
	// Arrange the views, resize ourselves
	float size = 2;
	NSRect bounds = [self bounds];
	
	NSEnumerator* viewEnum = [upperWindowViews objectEnumerator];
	ZoomSavePreview* view;
	
	while (view = [viewEnum nextObject]) {
		[view setFrame: NSMakeRect(0, size, bounds.size.width, 48)];
		size += 49;
	}
	
	NSRect frame = [self frame];
	frame.size.height = size;
	
	[self setFrameSize: frame.size];
	[self setNeedsDisplay: YES];
}

- (BOOL) isFlipped {
	return YES;
}

- (void) mouseDown: (NSEvent*) event {
}

- (void) previewMouseUp: (NSEvent*) evt
				 inView: (ZoomSavePreview*) view {
	int clicked = [upperWindowViews indexOfObjectIdenticalTo: view];
	
	if (clicked == NSNotFound) {
		NSLog(@"BUG: save preview not found");
		return;
	}
	
	if ([evt clickCount] == 1) {
		// Select a new view
		if (selected != NSNotFound) {
			[[upperWindowViews objectAtIndex: selected] setHighlighted: NO];
		}
		
		[view setHighlighted: YES];
		selected = clicked;
	} else if ([evt clickCount] == 2) {
		// Launch this game
		NSString* filename = [view filename];
		NSString* directory = [filename stringByDeletingLastPathComponent];
		
		if ([[[directory pathExtension] lowercaseString] isEqualToString: @"glksave"]) {
			// Pass off to the app delegate
			[[NSApp delegate] application: NSApp
								 openFile: directory];
		} else {
			//ZoomClient* newDoc = 
			[[NSDocumentController sharedDocumentController] openDocumentWithContentsOfFile: directory
																					display: YES];
		}
	}
}

- (NSString*) selectedSaveGame {
	if (selected >= 0 && selected != NSNotFound) {
		return [[upperWindowViews objectAtIndex: selected] filename];
	} else {
		return nil;
	}
}

- (BOOL) saveGamesAvailable {
	return saveGamesAvailable;
}

@end

File: zoom/src/zoomCocoa/ZoomSavePreviewView.swift
//
//  ZoomSavePreviewView.swift
//  Zoom
//
//  Created by C.W. Betts on 9/24/21.
//

import Cocoa
import ZoomView.ZoomUpperWindow
import ZoomPlugIns.ZoomStory

@objcMembers
class SavePreviewView: NSView {
	private var upperWindowViews = [SavePreview]()
	private var selected: Int? = nil
	private(set) var saveGamesAvailable: Bool = false

	override init(frame frameRect: NSRect) {
		super.init(frame: frameRect)
		autoresizesSubviews = true
		autoresizingMask = .width
	}
	
	required init?(coder: NSCoder) {
		super.init(coder: coder)
		autoresizesSubviews = true
		autoresizingMask = .width
	}
	
	override func draw(_ dirtyRect: NSRect) {
		// do nothing
	}
	
	func setDirectoryToUse(_ directory: String?) {
		// Get rid of our old views
		for view in upperWindowViews {
			view.removeFromSuperview()
		}
		upperWindowViews.removeAll()
		
		saveGamesAvailable = false
		selected = nil
		guard let directory = directory, FileManager.default.fileExists(atPath: directory) else {
			var ourFrame = self.frame
			ourFrame.size.height = 2
			self.frame = ourFrame
			needsDisplay = true
			return
		}
		// Get our frame size
		var ourFrame = self.frame
		ourFrame.size.height = 0
		
		// Load all the zoomSave files from the given directory
		let url = URL(fileURLWithPath: directory)
		guard let contents = try? FileManager.default.contentsOfDirectory(at: url, includingPropertiesForKeys: [.isDirectoryKey], options: [.skipsSubdirectoryDescendants, .skipsHiddenFiles]) else {
			return
		}

		for file in contents {
			switch file.pathExtension.lowercased() {
			case "zoomsave":
				let previewURL = file.appendingPathComponent("ZoomPreview.dat")
				
				var isDir: ObjCBool = false
				guard urlIsAvailable(previewURL, isDirectory: &isDir, isPackage: nil, isReadable: nil, error: nil), !isDir.boolValue else {
					continue
				}
				
				guard let preData = try? Data(contentsOf: previewURL) else {
					continue
				}
				var preVal: Any? = try? NSKeyedUnarchiver.unarchivedObject(ofClass: ZoomUpperWindow.self, from: preData)
				if preVal == nil {
					preVal = NSUnarchiver.unarchiveObject(with: preData)
				}
				
				guard let win = preVal as? ZoomUpperWindow else {
					continue
				}
				
				let preview = SavePreview(preview: win, with: previewURL)
				preview.autoresizingMask = .width
				preview.menu = self.menu
				addSubview(preview)
				upperWindowViews.append(preview)
				
				saveGamesAvailable = true
				
			case "glksave":
				let propertiesURL = file.appendingPathComponent("Info.plist")
				guard (try? propertiesURL.checkResourceIsReachable()) ?? false else {
					continue
				}
				
				guard let dat = try? Data(contentsOf: propertiesURL),
					  let previewProperties = try? PropertyListSerialization.propertyList(from: dat, options: [], format: nil) as? [String: Any] else {
					continue
				}
				
				guard let strID = previewProperties["ZoomGlkGameId"] as? String else {
					continue
				}
				
				let storyId = ZoomStoryID(idString: strID)
				let previewLinesURL = file.appendingPathComponent("Preview.plist")
				let previewLines: [Any]
				if (try? previewLinesURL.checkResourceIsReachable()) ?? false,
				   let dat2 = try? Data(contentsOf: previewLinesURL),
				   let previewLines2 = try? PropertyListSerialization.propertyList(from: dat2, options: [], format: nil) as? [Any] {
					previewLines = previewLines2
				} else {
					// Use some defaults if no lines are supplied
					if let story = (NSApp.delegate as! ZoomAppDelegate).findStory(storyId) {
						previewLines = ["Saved story from '\(story.title!)'"]
					} else {
						previewLines = ["Saved story"]
					}
				}
				
				let preview = SavePreview(previewStrings: previewLines, with: propertiesURL)
				preview.autoresizingMask = .width
				preview.menu = menu
				addSubview(preview)
				upperWindowViews.append(preview)
				
				saveGamesAvailable = true

			default:
				break
			}
		}
		
		// Arrange the views, resize ourselves
		var size: CGFloat = 2
		let bounds = self.bounds
		
		for view in upperWindowViews {
			view.frame = NSRect(x: 0, y: size, width: bounds.size.width, height: 48)
			size += 49
		}
		
		var frame = self.frame
		frame.size.height = size
		
		setFrameSize(frame.size)
		needsDisplay = true
	}
	
	@available(*, deprecated, renamed: "SavePreviewView.selectedSaveGameURL")
	var selectedSaveGame: String? {
		return selectedSaveGameURL?.path
	}
	
	var selectedSaveGameURL: URL? {
		if let selVal = selected {
			return upperWindowViews[selVal].fileURL
		} else {
			return nil
		}
	}

	
	func previewMouseUp(_ evt: NSEvent, in view: SavePreview) {
		guard let clicked = upperWindowViews.firstIndex(of: view) else {
			NSLog("BUG: save preview not found")
			return
		}
		
		if evt.clickCount == 1 {
			// Select a new view
			if let selected = selected {
				upperWindowViews[selected].isHighlighted = false
			}
			
			view.isHighlighted = true
			selected = clicked
		} else if evt.clickCount == 2 {
			// Launch this game
			let fileURL = view.fileURL!
			let directory = fileURL.deletingLastPathComponent()
			if directory.pathExtension.lowercased() == "glksave" {
				// Pass off to the app delegate
				_=NSApp.delegate?.application?(NSApp, openFile: fileURL.path)
			} else {
				NSDocumentController.shared.openDocument(withContentsOf: directory, display: true) { documen, documentWasAlreadyOpen, error in
					// do nothing
				}
			}
		}
	}
}
